src/dune_rules/pkg_rules.ml

```diff
@@ -1094,38 +1189,48 @@ end = struct
     | None -> Memo.return None
     | Some { Lock_dir.Pkg.build_command; install_command; depends; info; exported_env } ->
       assert (Package.Name.equal name info.name);
-      let* depends =
-        Memo.parallel_map depends ~f:(fun name ->
-          resolve db ctx name
-          >>| function
-          | `Inside_lock_dir pkg -> Some pkg
-          | `System_provided -> None)
-        >>| List.filter_opt
-      and+ files_dir =
-        let+ lock_dir = Lock_dir.get_path ctx >>| Option.value_exn in
-        Path.Build.append_source
-          (Context_name.build_dir ctx)
-          (Dune_pkg.Lock_dir.Pkg.files_dir info.name ~lock_dir)
-      in
-      let id = Pkg.Id.gen () in
-      let paths = Paths.make name ctx in
-      let t =
-        { Pkg.id
-        ; build_command
-        ; install_command
-        ; depends
-        ; paths
-        ; info
-        ; files_dir
-        ; exported_env = []
-        }
+      let* compiler =
```

Comment:
I agree with the comment above. Moreover, this behavior is actually dangerous because it makes the builds we compute from lock directories depend on non workspace related data. In particular, if a toolchain is installed that coincides with the name of a package in any existing lock directory, it will now be treated in a different way. If a package corresponds to a toolchain it should correspond to that in the lock directory.

I retract my approval until we see what can be done to address this issue.